openmls/src/group/core_group/new_from_welcome.rs

```diff
@@ -136,21 +122,56 @@ impl CoreGroup {
                 },
             };
 
-        // Commit secret is ignored when joining a group, since we already have
-        // the joiner_secret.
-        let (tree, _commit_secret_option) = TreeSync::from_nodes_with_secrets(
+        let tree = TreeSync::from_nodes(
             backend,
             ciphersuite,
             &nodes,
-            verifiable_group_info.signer(),
-            path_secret_option,
-            encryption_key_pair,
+            key_package_bundle
+                .key_package()
+                .leaf_node()
+                .encryption_key(),
         )
         .map_err(|e| match e {
             TreeSyncFromNodesError::LibraryError(e) => e.into(),
             TreeSyncFromNodesError::PublicTreeError(e) => WelcomeError::PublicTreeError(e),
         })?;
 
+        let diff = tree.empty_diff();
+
+        // Read the encryption key pair from the key store and delete it there.
+        let leaf_keypair = EncryptionKeyPair::read_from_key_store(
+            backend,
+            key_package_bundle.key_package.leaf_node().encryption_key(),
+        )
+        .ok_or(WelcomeError::NoMatchingEncryptionKey)?;
+        leaf_keypair
+            .delete_from_key_store(backend)
+            .map_err(|_| WelcomeError::NoMatchingEncryptionKey)?;
+
+        // If we got a path secret, derive the path (which also checks if the
+        // public keys match) and store the derived keys in the key store.
+        let group_keypairs = if let Some(path_secret) = path_secret_option {
+            let (path_keypairs, _commit_secret) = diff
+                .derive_path_secrets(
+                    backend,
+                    ciphersuite,
+                    path_secret,
+                    verifiable_group_info.signer(),
+                )
+                .map_err(|e| match e {
+                    TreeSyncSetPathError::LibraryError(e) => e.into(),
+                    TreeSyncSetPathError::PublicKeyMismatch => {
+                        WelcomeError::PublicTreeError(PublicTreeError::PublicKeyMismatch)
+                    }
+                })?;
+            vec![leaf_keypair]
+                .into_iter()
+                .chain(path_keypairs.into_iter())
+                .collect()
```

Comment:
It's a little unfortunate that we have to collect these. But let's keep this for now and only improve if we see that it's an issue.